### PR TITLE
Google Summer of Code

### PR DESCRIPTION
**Pull request by abhisharsinha**
This work was part of my Google Summer of Code project. The following are the major contributions of the project.

- Addition of multiple negatives ranking loss
- Addition of an example notebook for fine-tuning CLIP using multiple negatives ranking loss

### CODE REVIEW

**Comment by owenvallis** [comment]
Thanks for the PR! I'll review the notebook soon and I left a few comments on the loss.

**Comment by owenvallis** [comment]
@abhisharsinha, let me know your thoughts on the PR comments and we can get this merged into the dev branch.

File: tensorflow_similarity/losses/multinegrank_loss.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        distance: Which distance function to use to compute the pairwise.
+
+    Returns:
+        loss: loss value for the current batch.
+    """
+
+    pairwise_sim = distance(query_emb, key_emb) * scale
+    # considering similarity to be - distance
+    if distance.name != "inner_product":
+        pairwise_sim = -pairwise_sim
```
[the remark this answers is not in the document]
Comment (abhisharsinha) [reply tensorflow_similarity/losses/multinegrank_loss.py:52]: I agree negative distance is not an appropriate transformation for similarity in all cases. For now, I'll enforce inner_product.

**Review by owenvallis — CHANGES_REQUESTED** [review]

File: tensorflow_similarity/losses/multinegrank_loss.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    expd = tf.math.exp(pairwise_sim)
```
[the remark this answers is not in the document]
Comment (abhisharsinha) [reply tensorflow_similarity/losses/multinegrank_loss.py:53]: Thanks, I'll incorporate this and push the changes

**Comment by owenvallis** [comment]
Meant to revert the changes to `pyproject.toml` but deleted the file. Can you please add it back with the following

```
[tool.isort]
profile = "black"

[tool.black]
 line-length = 119
```

**Review by owenvallis — APPROVED** [review]

**Review by owenvallis — APPROVED** [review]

**Comment by owenvallis** [comment]
Looks like the Black and Isort formatting checks are failing for tensorflow_similarity/losses/multinegrank_loss.py. Otherwise looks good to go, thanks.

File: examples/multimodal_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,1708 @@
...
+      "id": "gIlFIhcappKp",
+      "metadata": {
+        "id": "gIlFIhcappKp"
+      },
+      "outputs": [],
+      "source": [
+        "# Installing dependencies\n",
+        "!pip install transformers\n",
+        "\n",
+        "!pip install git+https://github.com/abhisharsinha/similarity@gsoc"
```
Comment (owenvallis) [inline examples/multimodal_example.ipynb:130]: We'll want to update this once we merge into to the dev branch.

**Review by owenvallis — APPROVED** [review]
